### PR TITLE
LPS-115798 Escape html string from JSP before passing to the `message` property on openToast

### PR DESCRIPTION
**Pull request by diegonvs**
### Test Case:

Load a page in a different language from the default.
Like:
```
- Create a new site called Site Name with a public page called Home.
- Add an es_ES translation for Home page. eg. Casa.
- Navigate to the following url:
http://localhost:8080/es/web/site-name/home
```

### Description(Read while reading the code):

Generally I fixed a problem with escaping text before passing to `message` openToast parameter and I adjusted some code for cover all cases on these JSPs with openToast.

### What I did: 

* I noticed the HTML rendered from the `liferay-util:include` contains a lot of spaces causing an error on the browser, for solving this I wrapped the HTML string on a liferay-buffer for escaping this content for message props render this prop adequately.

* `user_locale_options.jsp` content is being imported via include taglib inside `top_message.jsp` taglib, a close button was being passed there and this button is unnecessary since it breaks with Lexicon standards and it can't call openToast callbacks because it was in a completely different context.

For solving this last point, I added an event on Liferay global event handler and for detecting when the openToast was closed. I consider this [LOC](https://github.com/liferay-frontend/liferay-portal/pull/39/files#diff-a86aa4def898981ade5704ca95af6e5eR49
) very dirty but I don't had any ideas to solve it adequately considering the limitations(rendering a html string inside a JSP which is using openToast). I would love your suggestions 💖 

Thanks!

### CODE REVIEW

**Comment by diegonvs** [comment]
ci:test:sf

File: portal-impl/src/com/liferay/portal/util/PortalImpl.java [hunk 1]
```diff
@@ -848,6 +848,8 @@ public void addUserLocaleOptionsMessage(
 
 		PortalMessages.add(
 			httpServletRequest, PortalMessages.KEY_ANIMATION, false);
+		PortalMessages.add(
```
Comment (jbalsas) [inline portal-impl/src/com/liferay/portal/util/PortalImpl.java:851]: I have real aversion and mistrust of anything that requires changing `PortalImpl`... specially for frontend. The less we touch here the better. Why is this even required?

File: portal-web/docroot/html/common/themes/top_messages.jsp [hunk 2]
```diff
@@ -28,31 +28,38 @@ String message = (String)PortalMessages.get(request, PortalMessages.KEY_MESSAGE)
 
 if (Validator.isNotNull(jspPath) || Validator.isNotNull(message)) {
 	String cssClass = GetterUtil.getString(PortalMessages.get(request, PortalMessages.KEY_CSS_CLASS), "alert-info");
+	String displayType = GetterUtil.getString(PortalMessages.get(request, PortalMessages.KEY_DISPLAY_TYPE), "success");
```
Comment (jbalsas) [inline portal-web/docroot/html/common/themes/top_messages.jsp:31]: Default value we set above in the `request` was `info` but now we default to `success`... I'm still failing to see where this is being set elsewhere in the code, so I'd still be reluctant to get this in.

@diegonvs, could you explain how is this supposed to be used from the outside?

**Comment by wincent** [comment]
> * contains a lot of spaces causing an error on the browser

Out of curiosity, what is the error?

I'm unfamiliar with `<liferay-util:buffer>`, but from the diff I gather that we were spewing a bunch of garbage into the middle of the `openToast` call, probably resulting in a syntax error. And I guess the buffer just accumulates the JSP output into a variable for later use, so that bit seems reasonable.

But I also don't see a clear connection between the buffering and the other changes, so looking forward to hearing the answer to @jbalsas questions.

**Comment by brunobasto** [comment]
`I'm unfamiliar with <liferay-util:buffer>, but from the diff I gather that we were spewing a bunch of garbage into the middle of the openToast call, probably resulting in a syntax error`

Yes, that's exactly what `<liferay-util:buffer>` does. Diego came to me with syntax error problems and I suggested using that. Looking at the diff now I'm not sure why it's need since it used to work before.. I was under the assumption the issue was with multiline strings

> The escaping makes sense, but the rest, setting an additional param in the request, firing an event and passing the jspPath makes me quite uneasy with this.

As I understand, the issue that https://github.com/liferay-frontend/liferay-portal/pull/39/files#diff-a86aa4def898981ade5704ca95af6e5eL31 used to put a close button with an `id` inside the HTML for the `message` prop. And then it would listen for an `onClick` on that button and do some stuff. Diego told me the toast already has a close button and that extraneous button was not working properly. I suggested firing an event to listed when a toast was closed instead. The problem I see is that https://github.com/liferay-frontend/liferay-portal/pull/39/files#diff-a86aa4def898981ade5704ca95af6e5eL52 needs to be executed only when a specific toast is closed and Diego is using `jspPath` to check for identity which is odd.

TL;DR He needs a way to listen when a specific toast was dismissed so the system can set some flags on don't bother the user again with that toast.

File: portal-impl/src/com/liferay/portal/util/PortalImpl.java (hunk 1 above)
Comment (diegonvs) [reply portal-impl/src/com/liferay/portal/util/PortalImpl.java:851]: If you debug the receiving values from `cssClass`, you will notice the way that people can customize the alert just passing another style there. However, the user couldn't override the alert Icons like exclamation, full-circle-check for each Alert. This fix avoids having an info Alert with a danger symbol e.g.

Screen Shot 2020-06-22 at 10 45 36

File: portal-web/docroot/html/common/themes/top_messages.jsp (hunk 2 above)
Comment (diegonvs) [reply portal-web/docroot/html/common/themes/top_messages.jsp:31]: Yes, default to success don't make any sense(considering https://github.com/liferay-frontend/liferay-portal/pull/39#discussion_r443570692), changing.

File: portal-impl/src/com/liferay/portal/util/PortalImpl.java (hunk 1 above)
Comment (jbalsas) [reply portal-impl/src/com/liferay/portal/util/PortalImpl.java:851]: Hey @diegonvs, I think I understand your point, but could you please put this in an simple example so we can all understand?

If it is what you mean, I think I'd be rather inclined to ignore this for now. I don't think this is an API we want to create and promote unless totally necessary. Did you find a specific use case where you can use this?

**Comment by diegonvs** [comment]
Regarding the `jspPath` match, I know this looks very weird but I didn't found other way to do this. I was trying to dig into PortalUtil like classes to see in what `request` can be used to providing a better match but I didn't found anything unique between these two JSPs.

File: portal-impl/src/com/liferay/portal/util/PortalImpl.java (hunk 1 above)
Comment (diegonvs) [reply portal-impl/src/com/liferay/portal/util/PortalImpl.java:851]: Because this part is responsible for rendering the user_locale_options JSP. https://github.com/liferay-frontend/liferay-portal/blob/e4e9201de2980247632bc2399b6649f9474c3c09/portal-impl/src/com/liferay/portal/util/PortalImpl.java#L855

Comment (jbalsas) [reply portal-impl/src/com/liferay/portal/util/PortalImpl.java:851]: So what happens if you don't make this change? Anything?

How would you use this from other place? Have you seen a locale_option message triggered from somewhere else with a different type?

Comment (diegonvs) [reply portal-impl/src/com/liferay/portal/util/PortalImpl.java:851]: I could don't touch on PortalImpl just matching the passed arguments to top_messages default values (and We're already doing this 08cfd05.

I'll remove my change on PortalImpl, since `displayType` is info by default.

**Comment by jbalsas** [comment]
Hey @diegonvs, please slow down, let's try not to make so many changes and try to talk through the issue and the possible solutions first!!

File: portal-web/docroot/html/common/themes/top_messages.jsp [hunk 3]
```diff
@@ -27,32 +27,39 @@ String jspPath = (String)PortalMessages.get(request, PortalMessages.KEY_JSP_PATH
...
-			message: "
-				<c:choose>
-					<c:when test="<%= Validator.isNotNull(jspPath) %>">
-						<liferay-util:include page="<%= jspPath %>" portletId="<%= portletId %>" />
-					</c:when>
-					<c:otherwise>
-						<liferay-ui:message key="<%= message %>" />
-					</c:otherwise>
-				</c:choose>",
+			message: '<%= HtmlUtil.escapeJS(alertMessage) %>',
```
Comment (jbalsas) [inline portal-web/docroot/html/common/themes/top_messages.jsp:51]: Not super fan about this, but maybe we could do here:

```jsp
<c:if test="<%= Validator.isNotNull(onCloseJspPath) %>">
onClose: function(event) {
    <liferay-util.include page="<%= onCloseJspPath %>" porltetId="<%= portletId %>" />
}
</c:if>
```

File: portal-web/docroot/html/common/themes/user_locale_options.jsp [hunk 4]
```diff
@@ -47,14 +45,11 @@ String currentURL = PortalUtil.getCurrentURL(request);
 	</div>
 
 	<aui:script use="aui-base,liferay-store">
-		var ignoreUserLocaleOptionsNode = A.one('#ignoreUserLocaleOptions');
-
-		ignoreUserLocaleOptionsNode.on(
-			'click',
-			function() {
+		Liferay.once('toastClosed', function (event) {
```
Comment (jbalsas) [inline portal-web/docroot/html/common/themes/user_locale_options.jsp:48]: If we did  https://github.com/liferay-frontend/liferay-portal/pull/39/files#r443635470, then we would "simply" need to extract this to a `user_locale_options_on_close.jsp`... which will look completely off for a JSP, but will avoid having to link this JSP through a global `toastClosed` API.

Comment (jbalsas) [reply portal-web/docroot/html/common/themes/user_locale_options.jsp:48]: @brunobasto, @diegonvs, @wincent... thoughts?

Comment (brunobasto) [reply portal-web/docroot/html/common/themes/user_locale_options.jsp:48]: You mean [here](https://github.com/liferay/liferay-portal/blob/f380f73537a0445f153e7a055fa36438b312bf69/portal-impl/src/com/liferay/portal/util/PortalImpl.java#L830) we would need to pass an additional attribute (`onCloseJspPath`) to the JSP? And probably add a new constant `PortalMessages.ON_CLOSE_JSP_PATH` to follow the pattern there? Feels dangerous. Those constants are exposed API and if we decide to take it out anytime it will be a major version semantic change.

I'd rather having a way to identify a toast and listen for events when it opens or closes. Something like 

```js
Liferay.openToast({
...
id / key: ''userLocaleToast'
...
})
```

But anyway we'd still need to pass some kind of value from https://github.com/liferay/liferay-portal/blob/f380f73537a0445f153e7a055fa36438b312bf69/portal-impl/src/com/liferay/portal/util/PortalImpl.java#L830 to the JSP. So maybe we really have to add another constant. Or, we could just set a request attribute maybe.

Comment (jbalsas) [reply portal-web/docroot/html/common/themes/user_locale_options.jsp:48]: > You mean here we would need to pass an additional message (onCloseJspPath) to the JSP? And probably add a new constant PortalMessages.ON_CLOSE_JSP_PATH to follow the pattern there? Feels dangerous. Those constants are exposed API and if we decide to take it out anytime it will be a major version semantic change.

Yeah... just trying to stay closer to what we have here... which is in fact just 2 instances. Only 2 places we use `PortalMessages.add` in the whole portal 😢

Comment (brunobasto) [reply portal-web/docroot/html/common/themes/user_locale_options.jsp:48]: Another though. What if we don't use the generic messages mechanism for this specific use case and make `user_locale_options.jsp` do its own call to `Liferay.openToast()` and in there it could add an `onClose` listener?

The tricky part would be to figure out how that JSP would be included on the page. I only found two usages of `PortalMessages.KEY_JSP_PATH` so maybe we can assume that whenever we need to include a jsp in PortalMessages, that jsp will be responsible for opening its own toast?

File: portal-web/docroot/html/common/themes/top_messages.jsp (hunk 3 above)
Comment (diegonvs) [reply portal-web/docroot/html/common/themes/top_messages.jsp:51]: So, inside this onCloseJSP We'll have just a script calling the things that needs to be called when closing, right?

File: portal-web/docroot/html/common/themes/user_locale_options.jsp (hunk 4 above)
Comment (wincent) [reply portal-web/docroot/html/common/themes/user_locale_options.jsp:48]: > I'd rather having a way to identify a toast and listen for events when it opens or closes.

We kind of have that already. You can pass a `containerId` when you call `openToast`, and @diegonvs has already added in this PR a `Liferay.fire` call that could be extended to include that in the callback data. (It is already passing `renderData`, which includes a `portletId`, which itself may be identifying enough, depending how many toasts a given portlet may render; and if it's not identifying enough, you can pass arbitrary data in `renderData`). Unless I am miscounting, that is three different ways to identify an individual toast to a greater or lesser degree.

Comment (jbalsas) [reply portal-web/docroot/html/common/themes/user_locale_options.jsp:48]: > I'd rather having a way to identify a toast and listen for events when it opens or closes. Something like

We've done this in some other places... `openToast` would do a `Liferay.component` call registering some type of proxy... and in this case that would mean wrapping the `onClose` prop to run it and fire an additional event so you could do `Liferay.componetnReady('id').on('close'...)`. 

While this is works, it favours a programming style that we're trying to run away from, so I'd want to defer the need for such APIs as much as we can.

Comment (jbalsas) [reply portal-web/docroot/html/common/themes/user_locale_options.jsp:48]: > The tricky part would be to figure out how that JSP would be included on the page. I only found two usages of PortalMessages.KEY_JSP_PATH so maybe we can assume that whenever we need to include a jsp in PortalMessages, that jsp will be responsible for opening its own toast?

Yeah, this is where this is inevitably leading. The more troubling issue is that this will likely cause a breaking change.

One option would be migrate the logic to a `TopHeadDynamicInclude` so we can control when it gets inserted and as you say, do the `openToast` call itself. Being in `portal-web`, though, not sure how much we can do or where we could move this to.

Comment (brunobasto) [reply portal-web/docroot/html/common/themes/user_locale_options.jsp:48]: So, what if we just leverage the `jspPath` being passed and just include it like the the snippet I drafted above? With that the breaking change would be the expected behavior of passing `PortalMessages.KEY_JSP_PATH `. Before it was assumed that we'd open the toast for them. With that change they'd have to open the toast themselves. It is a breaking change, but it doesn't look like passing `PortalMessages.KEY_JSP_PATH` is very popular.

Comment (jbalsas) [reply portal-web/docroot/html/common/themes/user_locale_options.jsp:48]: So, this is only used from `top_messages.jsp` as:

```
<%
if (!user.isDefaultUser() && !locale.equals(user.getLocale())) {
	PortalUtil.addUserLocaleOptionsMessage(request);
}
%>
```

So... I'd say we can move this out to an OSGi module with a `TopHeadDynamicInclude` and leave everything else as is, without expanding or breaking the existing API... as limited as it is :/

Comment (brunobasto) [reply portal-web/docroot/html/common/themes/user_locale_options.jsp:48]: Sounds reasonable. The other included JSP I found was https://github.com/liferay/liferay-portal/blob/master/portal-web/docroot/html/common/themes/layout_friendly_url_redirect.jsp but it doesn't do any JS so we're good there.

Comment (jbalsas) [reply portal-web/docroot/html/common/themes/user_locale_options.jsp:48]: Ok, asking for feedback on where to place this... I'll send a followup once that's clear because right now I honestly couldn't tell...

@diegonvs, I'll send a followup with some of the changes here but keeping the current API as stable as possible and we can continue conversing there, ok!?

Comment (diegonvs) [reply portal-web/docroot/html/common/themes/user_locale_options.jsp:48]: Ok @jbalsas!

**Comment by jbalsas** [comment]
See https://github.com/liferay-frontend/liferay-portal/pull/44 for an initial different approach!
